crystal_diffusion/models/mtp.py

```diff
@@ -0,0 +1,315 @@
+               (with roughly force < scale_by_force) get more weight. Defaults to 0.
+            bfgs_conv_tol (optional): Stop training if error dropped by a factor smaller than this
+                over 50 BFGS iterations. Defaults to 1e-3.
+            weighting (optional): How to weight configuration with different sizes relative to each other.
+                Choose from "vibrations", "molecules" and "structures". Defaults to "vibration".
+
+        Returns:
+            rc : return code of the mlp training script
+        """
+        train_structures, train_forces, train_stresses = check_structures_forces_stresses(
+            train_structures, train_forces, train_stresses
+        )
+        train_pool = pool_from(train_structures, train_energies, train_forces, train_stresses)
+        elements = sorted(set(itertools.chain(*[struct.species for struct in train_structures])))
+        self.elements = [str(element) for element in elements]  # TODO move to __init__
+
+        atoms_filename = "train.cfgs"
+
+        with ScratchDir("."):  # create a tmpdir - deleted afterew
+            atoms_filename = self.write_cfg(filename=atoms_filename, cfg_pool=train_pool)
+
+            if not unfitted_mtp:
+                raise RuntimeError("No specific parameter file provided.")
+            mtp_file_path = os.path.join(self.mlp_templates, unfitted_mtp)
+            shutil.copyfile(mtp_file_path, os.path.join(os.getcwd(), unfitted_mtp))
+            commands = [self.mlp_command, "mindist", atoms_filename]
+            with open("min_dist", "w") as f, subprocess.Popen(commands, stdout=f) as p:
+                p.communicate()[0]
+
+            # TODO check what min_dist is used for in maml
+            # with open("min_dist") as f:
+            #    lines = f.readlines()
+            # split_symbol = "="  # different for mlip-2 (":") and mlip-3 ("=")
+            # min_dist = float(lines[-1].split(split_symbol)[1])
+
+            save_fitted_mtp = ".".join([unfitted_mtp.split(".")[0] + "_fitted", unfitted_mtp.split(".")[1]])
+            cmds_list = [
+                self.mlp_command,
+                "train",
+                unfitted_mtp,
+                atoms_filename,
+                f"--save_to={save_fitted_mtp}",
+                f"--iteration_limit={max_iter}",
+                "--al_mode=nbh",  # active learning mode - required to get extrapolation grade
+                # f"--curr-pot-name={unfitted_mtp}",  # TODO check those kwargs
+                # f"--energy-weight={energy_weight}",
+                # f"--force-weight={force_weight}",
+                # f"--stress-weight={stress_weight}",
+                # f"--init-params={init_params}",
+                # f"--scale-by-force={scale_by_force}",
+                # f"--bfgs-conv-tol={bfgs_conv_tol}",
+                # f"--weighting={weighting}",
+            ]
+            with subprocess.Popen(cmds_list, stdout=subprocess.PIPE) as p:
```

Comment:
This `Popen` could be hidden away in a `_method`.

Reply:
should be hidden now